### PR TITLE
Hardcode the project version

### PR DESCRIPTION
**Pull request by Gliese852**
Since the build version is actually the date the build was run, if it is rebuilt on a different day it becomes erroneous. For example, since in any distribution the build into the image will most likely occur after the release, or even rebuild it regularly, the version will differ from the release one, which can be misleading.

There is also a second part of the version - the commit hash, but if the stable version is (re)built from the source archive (which is correct), there is no git repo at all.

Therefore, it is proposed to freeze this manually (with a script) in CMakeLists, so that it unambiguously indicates the version of the game.

### CODE REVIEW

**Comment by impaktor** [comment]
Ping @pcercuei 
(this came up with people playing the new flatpak build)

**Comment by pcercuei** [comment]
I disagree about hardcoding the values. This is the worst solution.

What I'd suggest - get the latest tag that was created, which is always the build date.
```
git describe --abbrev=0 --match 20*
```

Of course, make sure that this tag pattern is reserved for releases.

**Comment by pcercuei** [comment]
Or, since you don't like to annotate your release tags,
```
git describe --tags --match 20*
```

**Comment by Gliese852** [comment]
@pcercuei the problem is that, as already written, the original git repo is not always available. For example, in a flatpak, do you see this tag when building?

**Comment by Gliese852** [comment]
I would like to add that, as far as I know, Fedora, OpenSUSE, and ALT Linux build Pioneer from the source archive and will not see this tag. Moreover, no one guarantees that any service tag in the internal package repository will not match the proposed glob.

**Comment by Gliese852** [comment]
Hmm, the [screenshot](https://github.com/pioneerspacesim/pioneer/issues/5712#issuecomment-1890532884) from the guy who uses the flatpak version, you can clearly see that there is no git repo during the build (no hash).

Also I thought, it seems it would be better to hardcode only the "date-version", but get (or not) the commit hash during the build. This will give us more information about the origins of a particular version and may prevent some unpleasant surprises.

**Comment by impaktor** [comment]
This change is of no use if package maintainers don't know about it. Therefore: might I suggest a section in README.txt, called "Package maintainers", or "Versioning packages", or similar, that explains where/how to get the appropriate version?

File: CMakeLists.txt
Comment (Web-eWorks) [inline CMakeLists.txt:121]:
```suggestion
set(PROJECT_VERSION 20240113
	CACHE STRING "Version identifier for current compiled build")

if (NOT PROJECT_VERSION_GIT)
	string(TIMESTAMP PROJECT_VERSION_GIT "built %Y%m%d")
endif()
```

**Review by Web-eWorks — CHANGES_REQUESTED** [review]
This is a good idea in theory - in practice there are a few additional considerations:

1. The version number needs to be correct "out of the box" for distribution packaging specifically. We cannot rely on every package maintainer knowing the correct invocation to ensure a consistent version number is displayed on the main screen. This is not a concern for "single user" builds, as presumably if a user is making a bug report from their own source build, they're technically competent enough to correctly report the commit or branch they're building from.
2. We need some way to clearly indicate non-release builds as such, especially when it comes to CI builds (where we may want to e.g. version a build as `devel-pr#5600 (f19c0a13)`). Any version string should be configurable via CMake command line parameter.
3. Ideally, in any situation where a git repo is available, the secondary build information should display the git commit hash and/or specific platform/build configuration set at build time. E.g. `20240203 (f129104912 x64+avx2)`. If not available, a fallback to the build timestamp may be appropriate.

I've left a suggested change that would allow specifying custom build versions for our CI workflows, as well as falling back to timestamp if there isn't an explicitly set version info string.

I don't think there's a need for an `update-version` script at all, given that we will only be modifying the value defined in CMakeLists.txt when tagging a new release (and most likely in the tagged commit which constitutes that release.)

I'd suggest it might be a good idea to move the definitions of `PROJECT_VERSION` and `PROJECT_VERSION_GIT` all the way to the top of CMakeLists.txt for visibility. This would make it easier to append to the `PROJECT_VERSION_GIT` variable (really, the build info string) with specific build options as they're defined.

**Comment by Web-eWorks** [comment]
> Or maybe we can go a little further and generalize PIONEER_VERSION_GIT to PIONEER_VERSION_AUX or something like that? 

I'd recommend adding a new variable `PIONEER_VERSION_INFO` which has the git commit information appended to it if present (and also specific build types as needed).

**Comment by Gliese852** [comment]
@Web-eWorks 

> 1. The version number needs to be correct "out of the box" for distribution packaging specifically. We cannot rely on every package maintainer knowing the correct invocation to ensure a consistent version number is displayed on the main screen. This is not a concern for "single user" builds, as presumably if a user is making a bug report from their own source build, they're technically competent enough to correctly report the commit or branch they're building from.

If the maintainer does not specify anything, it will be just a version (without additional information in brackets). The absence of an additional version will already tell us that this is just a release archive, and there was no git repository. If the maintainer also additionally indicates useful information in brackets, thank him, but this is not critical.

> 3\. Ideally, in any situation where a git repo is available, the secondary build information should display the git commit hash and/or specific platform/build configuration set at build time. E.g. `20240203 (f129104912 x64+avx2)`. If not available, a fallback to the build timestamp may be appropriate.

That is, by default, it will be like this: `20240203 (built 20241010)` but is it worth it to overload the version string like that? What does the build date tell us?

> I don't think there's a need for an `update-version` script at all, given that we will only be modifying the value defined in CMakeLists.txt when tagging a new release, and specifically setting that value to the tag name for the release (and most likely in the tagged commit which constitutes that release.)

Why not update this more often? For those who build unstable packages, for example. If they also have their own version history, there will be a timestamp of the latest release (e.g. half a year ago) and some hash unknown to us. We'll have to suffer again.

**Comment by Web-eWorks** [comment]
> That is, by default, it will be like this: 20240203 (built 20241010) but is it worth it to overload the version string like that? What does the build date tell us?

On second thought, the build timestamp is probably not as useful as I was thinking when I originally proposed that idea. I'd like to keep the commit hash  + build flags however in any case where that information is available.

> Why not update this more often? For those who build unstable packages, for example. If they also have their own version history, there will be a timestamp of the latest release (e.g. half a year ago) and some hash unknown to us. We'll have to suffer again.

I don't see a need to do this, based on our current and past volume of bug reports.

All of our bug-reporting traffic on unstable versions of Pioneer in the last year that I'm aware of came from (in order of volume):

1. self-compiled checkouts of the game code, which have valid git hashes (unless they're actively making commits and developing the game);
2. users who've installed `pioneer-git` from the AUR, which has valid git hashes because it's a direct compilation of our sources without patches;
3. users downloading our CI build artifacts, which will have extremely detailed version information set directly from our CI workflow script.

To the best of my knowledge there are no distributions with an unstable build of Pioneer other than Arch's `pioneer-git` which are responsible for more than 1-3 bug reports _per year_. For the above reasons, I don't see any need to modify the "stable version number" on our end outside of officially-tagged pre-releases / release candidate builds.

 I think there's a danger here of trying to "be all things to all people" in a way that will consume far more of our developer time than it would save in terms of bug report clarity. For a hardcoded "unstable version number" to be effective, we would have to be changing the version number after every single PR merge or commit made to the master branch.

At most, I'd suggest authoring one commit per release immediately after the tagged commit which changes `PROJECT_VERSION` to something like `"20240203-develop"` - that would indicate it's an unstable version and provide some general timeframe for what commit the binary is built from.

**Comment by Gliese852** [comment]
Yes, we should probably just release more often, beta releases or something like that.

File: CMakeLists.txt
Comment (Gliese852) [reply CMakeLists.txt]: Done, except for the default timestamp.

File: CMakeLists.txt
Comment (Web-eWorks) [inline CMakeLists.txt:8]:
```suggestion
set(PROJECT_VERSION 20240119-develop
	CACHE STRING "Version identifier for current compiled build")
```

**Review by Web-eWorks — APPROVED** [review]
I'd recommend adding the `-develop` suffix for non-release builds, but this looks good. Once squashed and merged I'll look at modifying the CI workflows to annotate with build information.

File: CMakeLists.txt
Comment (Gliese852) [reply CMakeLists.txt]: Made a version as discussed later.

**Comment by Gliese852** [comment]
Forced it all together into one clean commit.

**Comment by impaktor** [comment]
@pcercuei We just published a new release, do you have time to fix a new flatpak for us, pretty please?

**Pull request merged**
